# German Ebay Car Sales Analysis

## Data Source and Intro

The following dataset is a subset of what was provided on Kaggle at the following url below. A subset of 50,000 data points will be analyzed.
https://www.kaggle.com/orgesleka/used-cars-database/data

For the purpose of this guided mission from Dataquest.io, some of the data has been "dirtied" in order to practice cleaning up data for analysis. The data dictionary provided by the Dataquest.io guided mission is below:


- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the - car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

## Questions to Analyze

Going into this dataset, I hope to look into the following:

- What was the most popular brand of car by listing?
- How effective were sales for cars that had listings included in an A/B test versus those that did not?
- What is the average price based on how many kilometers that car had been driven?

## Methods/Tools Used for Analysis

I am going to use the NumPy and pandas libraries to clean the data for analysis.

In [57]:
# import the pandas and NumPy libraries
import numpy as np
import pandas as pd

In [58]:
# read the autos.csv file into pandas
autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [59]:
# test to see what the dataset looks like with the above selected encoding
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

In [60]:
# read the autos.csv file into pandas using a different encoding to compare
autos = pd.read_csv("autos.csv",encoding="Windows-1252")

In [61]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere 

In [62]:
# the above datasets look the same, so I'll just use Latin-1 for now
# take a look at the info of the dataframe
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## Initial Analysis of Info() on Autos Dataset

Below are some initial thoughts on looking at the dataset:

- The variable types for all of the columns are either int64 or object. For some, these look to be as expected at first glance - yearOfRegistration has to be an integer since you can't have a fraction of a year, so it is rightfully an int64. name is an object, which seems okay to have somethting that would normally be a string as an object. However, there are other columns that need to be casted as a different variable type for easier analysis such as dateCrawled and dateCreated needing to be Date values, and price should be a float
- It appears that there are some null values for certain fields, which is to be expected (especially with a "dirtied" version of dataset). However, there is one field that I think is kind of alarming to have as a null field at all - notRepairedDamage. Roughly 20% of the listings do not have a value filled in for this. One could make an assumption that if the owner of the listing did not list a value for this field that there is no damage; however, I'm not sure how what kind of laws Germany has in place to protect consumers. I'm sure there are some guidelines that eBay has in general to help protect consumers, but again, I'm not sure if this is a required field or not for owners to fill out. It most likely is not since there are null fields out there. If we were to fill in the nulls in this data, I'd want to err on the side of caution and go ahead and put a value for "Assume Damage - inquire from seller" or something along the lines.

In [63]:
# look at the first five records of the data set to see if there is anything we can glean from the data
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Analysis of first five records of dataset

From looking at the first five records of the dataset, I had the following initial observations:

- The entries are in German! This shouldn't be a surprise since the listings are coming from a German eBay, and I'm exposing my ignorance and ethnocentrism in thinking that everything should be in the English language. Even though my knowledge of German is slim, I can for the most part make out that "privat" refers to a private seller in the seller column, and I know "nein" means no in the notRepairedDamage column from Quentin Tarantino's film Ingluourious Basterds. If I am at a loss, I'll try my hand with Google Translate.
- The site being a German site, I still see dollar symbols on the price. I wonder if this is referring to US dollars or Euros? I guess it doesn't matter terribly much if all of the entries are in US dollars, but I have a feeling there will be multiple types of currency used, and we'll need to convert to a single currency in order to do thorough analysis on that column.
- I noticed at least on the first five entries that there are 0 pictures for each listing. That worries me because if someone is selling a car, I'd like to see a visual of what I am going to buy! I'll be interested in seeing how long a listing has been on eBay without having a buyer and how many pictures that listing has.

In [64]:
# let's rename the columns to snake case instead of camel case
# first, let's look at the columns again
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [65]:
# use the rename function that is associated with the pandas dataframe to rename the columns
autos = autos.rename(columns = {
    'dateCrawled':'date_crawled',
    'offerType':'offer_type',
    'abtest':'ab_test',
    'vehicleType':'vehicle_type',
    'yearOfRegistration':'registration_year',
    'powerPS':'power_ps',
    'monthOfRegistration':'registration_month',
    'fuelType':'fuel_type',
    'notRepairedDamage':'unrepaired_damage',
    'dateCreated':'ad_created',
    'nrOfPictures':'nr_of_pictures',
    'postalCode':'postal_code',
    'lastSeen':'last_seen'
})


## Reasoning for Renaming the Columns

You may have noticed that we went ahead and renamed the columns to snakecase rather than camelcase. Ultimately it may come down to preference, but it just helps to have all of the columns in a uniform format so that when you call a certain column, you're not sitting there thinking, "Now wait, was there a capital letter in this column name?" With snakecase, there is never a capital letter, and any spaces yield an underscore. Now that the columns are in snakecase form, I don't plan on having to keep calling dataframe.head() just to double check that I got the column right!

In [66]:
# now let's see if we can clean some of the data in the dataset
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [67]:
# let's look at the following sample of records to see what some data look like
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Data that needs to be cleaned

Let's look at the following categories for cleaning the data:

- Any columns that have mostly one value that are candidates to be dropped
- Any columns that need more investigation
- Any examples of numeric data stored as text that needs to be cleaned

We'll see if we can find examples of each of the categories above below:

## Any columns that have mostly one value that are candidates to be dropped

Both of the below columns only have two unique values, and one of those values appears 49999 times out of 50000. I think it's safe to say since 99.99% of the data behaves this way, I don't believe we will find anything earth shattering from the one record that doesn't conform to this. With that, we can drop these columns:

- seller
- offer_type

## Any columns that need more investigation

At first glance, I'm not able to discern whether this column needs to be cleaned up or if it's ready for use or if I want to use it at all. I need to look into these more after I take care of cleaning the low-hanging fruit:

- vehicle_type: Mainly because I don't know German and need to figure out what translates to "truck," "car," etc. in order to dig deeper into this column if I want to
- power_ps: Not sure what this means, so I need to do some research on this field to actually understand what it is trying to say
- fuel_type: Same as vehicle_type above. I think it may be helpful to know what kind of gas these vehicles take as that will be part of the cost to maintain the car
- nr_of_pictures: I'm curious if this column only shows 0's. If so, I can drop this column as well. I think I just need to make it an integer first before I analyze it.

## Any examples of numeric data stored as text that needs to be cleaned.

Below are some columns that appear to be stored as text but need to be numeric:

- date_crawled: Change it to a date
- price: take off the money sign and make it a float
- registration_year: make this an integer
- power_ps: make this a float?
- odometer: take off the km and make this a float?
- month_of_registration: make this an integer
- nr_of_pictures: make this an integer
- postal_code: make this an integer
- last_seen: make this a date

Now that we've laid out which columns to look at, let's first wrap up those that we planned on investigating.

In [68]:
# let's look up the unique values for vehicle_type. There should be 8
autos["vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [69]:
# now that we have a list of values in German, let's look up what they mean in English rename the values to English values for easier analysis
car_name_mapping = {
    "limousine":"sedan",
    "kleinwagen":"small car",
    "kombi":"van",
    "bus":"bus",
    "cabrio":"convertible",
    "coupe":"sport sedan",
    "suv":"suv",
    "andere":"other"
}

autos["vehicle_type"] = autos["vehicle_type"].map(car_name_mapping)

In [70]:
# now check the values to see what we have now
autos["vehicle_type"].value_counts()

sedan          12859
small car      10822
van             9127
bus             4093
convertible     3061
sport sedan     2537
suv             1986
other            420
Name: vehicle_type, dtype: int64

In [71]:
# let's look and see what the different fuel_type values are
autos["fuel_type"].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [72]:
# similar to fuel_type, let's translate!
fuel_name_mapping = {
    "benzin":"gasoline",
    "diesel":"diesel",
    "lpg":"liquified petroleum gas",
    "cng":"compressed natural gas",
    "hybrid":"hybrid",
    "andere":"other",
    "elektro":"electric"
}

autos["fuel_type"] = autos["fuel_type"].map(fuel_name_mapping)

In [73]:
# now check the values to see what we have now
autos["fuel_type"].value_counts()

gasoline                   30107
diesel                     14567
liquified petroleum gas      691
compressed natural gas        75
hybrid                        37
other                         22
electric                      19
Name: fuel_type, dtype: int64

In [74]:
# now, let's check the number of pictures
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

## Findings on Columns I wanted to investigate further

- vehicle_type: it looks like we have 420 noted as "other." Let's just hope that there's some sort of description for this vehicle with some picutres or else I for sure wouldn't buy if I didn't know what it was!

- power_ps: it turns out that ps stands for "Pferdestrke," which literally translates to "horse strength" (https://www.whatcar.com/news/5-what-does-ps-and-bhp-mean/n12325). So, this measure translates to horse power!

- fuel_type: it looks like most of the fuel types translate over to English okay. There is just a large general "gasoline" category though that I thought would be more broken up into premium, unleaded, etc. I don't really understand cars to be honest, but I'm learning by going through all of this, and given how general this category is with being largely gasoline or diesel, I don't think I need to investigate this category further.

- nr_of_pictures: it looks like unfortunately there are no pictures for any of these listings. That concerns me as I feel that havings pictures of what you are trying to sell (especially something that is a heavy investment like a vehicle), your potential buyers would want to see what they could be potentially buying! Perhaps, since these are open listings, these cars have been sitting out there because they have no pictures? Or, maybe for this dataset, the curator decided to leave out those that had pictures to make the dataset smaller (and maybe the pictures were included in the actual dataset somehow?). Regardless, I can only speculate, and there's nothing to glean from this column with no data. As a result, this column can be dropped as well.

In [75]:
# now let's get to cleaning some data! Let's start with odometer
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [76]:
# let's take out the km from each of these values and make them an int (even though they could be float, all of the values are whole numbers, so I'm making them an int)
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [77]:
# now let's rename the column to odometer_km
autos = autos.rename(columns={"odometer":"odometer_km"})

In [78]:
# now we just need to make the price into the proper format
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [79]:
# now let's rename the column to price_dollars
autos = autos.rename(columns={"price":"price_dollars"})

In [80]:
# now we'll analyze the odometer_km and price columns to see how a sample of the data looks. Let's start with odometer_km
autos["odometer_km"].unique().shape

(13,)

In [81]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [82]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [83]:
# now we'll analyze the price_dollars
autos["price_dollars"].unique().shape

(2357,)

In [84]:
autos["price_dollars"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [85]:
autos["price_dollars"].value_counts().sort_index(ascending=True)

0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
190000.0         1
194000.0         1
197000.0         1
198000.0         1
220000.0         1
250000.0         1
259000.0         1
265000.0         1
295000.0         1
299000.0         1
345000.0         1
350000.0         1
999990.0         1
999999.0    

## On Cleaning Up odometer_km and price_dollars

### odometer_km

When looking at a sample of values for odometer_km, the values look reasonable. There are some that were driven a bunch past 150,000 km, which roughly translates to a little over 93,000 miles. That's a reasonable amount on an odometer I think as I'm getting close to that myself (and my car is still trucking along like a champ! Nope I don't own a truck).

### price_dollars

When looking at price_dollars, however, I noticed a few things:

- There were 1421 entries with 0 dollars. Why even try to post a car on eBay if you're not expecting any money from it? I think this is bogus data, so I'm going to remove these records.

- There's also one entry with a price tag of 99999999 dollars. Maybe the seller wanted to try their luck and put int he most expensive amount? Either way, I don't think it's reasonable, and it throws off the average "reasonable" price, so I'm going to remove this amount as well.


In [86]:
# remove the 0 records from price_dollars
autos = autos[autos["price_dollars"] > 0]

In [87]:
# verify that the 0's are gone now
autos["price_dollars"].value_counts().sort_index(ascending=True)

1.0           156
2.0             3
3.0             1
5.0             2
8.0             1
9.0             1
10.0            7
11.0            2
12.0            3
13.0            2
14.0            1
15.0            2
17.0            3
18.0            1
20.0            4
25.0            5
29.0            1
30.0            7
35.0            1
40.0            6
45.0            4
47.0            1
49.0            4
50.0           49
55.0            2
59.0            1
60.0            9
65.0            5
66.0            1
70.0           10
             ... 
151990.0        1
155000.0        1
163500.0        1
163991.0        1
169000.0        1
169999.0        1
175000.0        1
180000.0        1
190000.0        1
194000.0        1
197000.0        1
198000.0        1
220000.0        1
250000.0        1
259000.0        1
265000.0        1
295000.0        1
299000.0        1
345000.0        1
350000.0        1
999990.0        1
999999.0        2
1234566.0       1
1300000.0       1
3890000.0 

In [88]:
# Now remove the 99999999 record
autos = autos[autos["price_dollars"] < 99999999]

In [89]:
# verify that the record is no longer there:
autos["price_dollars"].value_counts().sort_index(ascending=True)

1.0           156
2.0             3
3.0             1
5.0             2
8.0             1
9.0             1
10.0            7
11.0            2
12.0            3
13.0            2
14.0            1
15.0            2
17.0            3
18.0            1
20.0            4
25.0            5
29.0            1
30.0            7
35.0            1
40.0            6
45.0            4
47.0            1
49.0            4
50.0           49
55.0            2
59.0            1
60.0            9
65.0            5
66.0            1
70.0           10
             ... 
145000.0        1
151990.0        1
155000.0        1
163500.0        1
163991.0        1
169000.0        1
169999.0        1
175000.0        1
180000.0        1
190000.0        1
194000.0        1
197000.0        1
198000.0        1
220000.0        1
250000.0        1
259000.0        1
265000.0        1
295000.0        1
299000.0        1
345000.0        1
350000.0        1
999990.0        1
999999.0        2
1234566.0       1
1300000.0 

In [90]:
autos[["date_crawled","ad_created","last_seen","registration_year","registration_month"]][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  registration_month  
0               2004                   3  
1               1997                   6  
2               2009                   7  
3               2007                   6  
4               2003                   7

In [91]:
# now let's figure out the distribution of days for date_crawled
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64

### On the date_crawled distribution

It appears that the data was gathered pretty evenly across the beginning of March until the beginning of April. There's not much that I'm gleaning from this, so I'll move on to analyzing the next date column.

In [92]:
# Now, let's look the last_seen column:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: last_seen, dtype: float64

### On the last_seen column

Now that's interesting. Whereas the date_crawled column was spread pretty evenly, the last_seen column has more dates later on in the dataset. However, thinking about it more, this makes sense because the last_seen column I assume is taking the max date of when the car listing was viewed. As a result, even if a car listing was looked at every day, only the latest date would appear. The only thing I wonder about the data now though is if the listings were updated frequently, would the crawler pick up the latest information or only once? That's a question I won't answer on this round, but that's something I'll consider for future data analyses.

In [93]:
# Now, let's look the ad_created column:
autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033163
2016-03-10    0.031887
2016-03-11    0.032896
2016-03-12    0.036766
2016-03-13    0.017004
2016-03-14    0.035181
2016-03-15    0.034007
2016-03-16    0.030117
2016-03-17    0.031290
2016-03-18    0.013586
2016-03-19    0.033678
2016-03-20    0.037939
2016-03-21 

### On the ad_created column

It appears that most of the ads were created in the month of March and into the beginning of April.

In [94]:
# Let's look at the distribution of the registration_year values
autos["registration_year"].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### On the registration_year column

Below are some of my observations on the distribution of registration_year values:

- It appears that the registration_year is null for roughly 1500 listings. Are these vehicles not registered at all, or did the creator of the listing just to forget to put the registration_year? If I were interested in one of these cars, I would reach out to the creator of the listing to clarify this fact.
- There is a minimum value of 1000, as there is at least one listing with a car that was reportedly registered in the year 1000. Automobiles didn't come around until the early 1900s, so unless it was a Roman chariot for sale, there's no way that this is correct. I would remove this record from the dataset.
- There is also a registration year of 9999 for at least one listing. Unless this was a time-traveling vehicle, there is no way that a car could be registered for a year in the future. I would remove this record from the dataset as well.
- It looks like the average registration_year was 2004 (that's including the 1000 and 9999 years though as a headsup).

In [95]:
# Make sure that registration_year is an int
autos["registration_year"].astype(int)

0        2004
1        1997
2        2009
3        2007
4        2003
5        2006
6        1995
7        1998
8        2000
9        1997
10       2017
11       2000
12       2010
13       1999
14       2007
15       1982
16       1999
17       1990
18       1995
19       2004
20       2003
21       2004
22       2015
23       2010
24       2014
25       1996
26       1992
28       2007
29       2004
30       2002
         ... 
49968    1986
49969    2005
49970    2010
49971    2001
49972    2004
49973    2004
49975    2012
49976    1992
49977    2003
49978    1996
49979    2011
49980    1995
49981    1998
49982    2004
49983    1999
49985    1995
49986    2010
49987    2013
49988    2001
49989    1997
49990    2012
49991    2016
49992    2009
49993    1997
49994    2001
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48578, dtype: int64

In [96]:
# now, do a value counts of what we are seeing for the registration_year
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      24
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2637
2002    2486
2003    2699
2004    2703
2005    2936
2006    2671
2007    2277
2008    2215
2009    2086
2010    1589
2011    1623
2012    1310
2013     803
2014     664
2015     392
2016    1220
2017    1393
2018     471
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

In [97]:
# since the data was collected in 2016, let's only pull those that are less than 2016
autos = autos[autos["registration_year"] <= 2016]

In [98]:
# from there, since, according to wikipedia, in 1908, the Model T was the first widely available automobile to the masses, we'll look at 1908 on.
autos = autos[autos["registration_year"] >= 1900]

In [99]:
# now, let's look at what the average registration year is
autos["registration_year"].describe()

count    46692.000000
mean      2002.908250
std          7.190078
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

### On cleaning up the registration_year field

I performed the following clean-up activities on the registration_year field:

- Since the data was collected on 2016, I removed records that were registered pasted the year 2016 since a car cannot be registered on a future date.
- Since the Model T was the first vehicle that was available to the masses started in 1908, I removed records where the vehicle had a registration year before the year 1900

In [100]:
# Now, let's calculate the distribution of the remaining values:
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000514
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002891
1989    0.003727
1990    0.007432
1991    0.007260
1992    0.007946
1993    0.009102
1994    0.013471
1995    0.026300
1996    0.029405
1997    0.041784
1998    0.050608
1999    0.062066
2000    0.067592
2001    0.056476
2002    0.053243
2003    0.057804
2004    0.057890
2005    0.062880
2006    0.057205
2007    0.048766
2008    0.047439
2009    0.044676
2010    0.034032
2011    0.034760
2012    0.028056
2013    0.017198
2014    0.0142

### On looking at the new distribution of the registration_year values:

it looks like the majority of vehicles were registered between 1994 and 2016.

In [101]:
# Now, let's look at brands and what the average price associated with some brands is
# first, let's look at the different kind of brands
autos["brand"].value_counts(normalize=True).index[0:5]

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')

### On which brands I chose

I chose the following top 5 brands based on highest amount of vehicles listed in the dataset:

- volkswagen
- bmw
- opel
- mercedez_benz
- audi

In [102]:
# now, let's create an empty dictionary to hold the aggregate data
top_five_brands_average_price = {}

# then, let's loop through and get the average price per those top five brands
# first, let's get the index list to loop through
top_five_brands = autos["brand"].value_counts(normalize=True).index[0:5]

# then, let's start the loop
for i in top_five_brands:
    # then, set the selected brand to a variable
    selected_brand = autos[autos["brand"] == i]
    # from there, find the mean of the price for that brand
    mean_price = selected_brand["price_dollars"].mean()
    # finally, make the brand in the dictionary associated with that mean price
    top_five_brands_average_price[i] = mean_price
    
# print the dictionary to see the average price for those top 5 brands
print(top_five_brands_average_price)

{'volkswagen': 6729.81956411556, 'audi': 9336.687453600594, 'mercedes_benz': 8628.450366422385, 'bmw': 8571.480147917478, 'opel': 5432.479195699781}


### On the mean price per the top five brands

From looking at the mean price per the top five brands, the following brands are ordered by most expensive to least expensive of the top five:

- audi: 9336.69 dollars
- mercedez_benz: 8628.45 dollars
- bmw: 8571.48 dollars
- volkswagen: 6729.82 dollars
- opel: 5432.48 dollars

In [103]:
# now let's do the same for mean mileage that we did for mean price:
# now, let's create an empty dictionary to hold the aggregate data
top_five_brands_average_mileage = {}

# then, let's loop through and get the average price per those top five brands
# first, let's get the index list to loop through
top_five_brands = autos["brand"].value_counts(normalize=True).index[0:5]

# then, let's start the loop
for i in top_five_brands:
    # then, set the selected brand to a variable
    selected_brand = autos[autos["brand"] == i]
    # from there, find the mean of the price for that brand
    mean_mileage = selected_brand["odometer_km"].mean()
    # finally, make the brand in the dictionary associated with that mean price
    top_five_brands_average_mileage[i] = mean_mileage
    
# print the dictionary to see the average price for those top 5 brands
print(top_five_brands_average_mileage)

{'volkswagen': 128713.6340598074, 'audi': 129157.38678544914, 'mercedes_benz': 130788.36331334666, 'bmw': 132575.9050214091, 'opel': 129314.15488751742}


In [105]:
# now, create a new series for the first price dictionary
bmp_series = pd.Series(top_five_brands_average_price)

# from there, make the series a dataframe to add the mean milelage to it
df = pd.DataFrame(bmp_series, columns=['mean_price'])

In [107]:
# create a series for the average mileage
brands_mean_mileage = pd.Series(top_five_brands_average_mileage)

In [109]:
df["mean_mileage"] = brands_mean_mileage

In [110]:
df

mean_price   mean_mileage
audi           9336.687454  129157.386785
bmw            8571.480148  132575.905021
mercedes_benz  8628.450366  130788.363313
opel           5432.479196  129314.154888
volkswagen     6729.819564  128713.634060

### On mean price and mean "mileage" of top five brands

When we are looking at mileage, we are looking at what the dataset has available - how much the car has been driven, and in this case, the odometer_km column. Since, for the top five brands, the odometer_km mean values are roughly the same, then there might be something else that influences makes a particular brand more expensive.

# So long for now

There are some next steps to look into, but I am going to hang my hat on this project for now. Thank you for your time, and until the next one!